### ISSUE TITLE
Make connector-test-kit run with indexes on mongodb

### ISSUE THREAD

**Issue by tomhoule** [prisma/prisma-engines#2361]
TL;DR : the migration engine `SchemaPush` command was implemented for MongoDB in https://github.com/prisma/prisma-engines/pull/2287

It is the command that is used by the query engine to set up databases for tests in connector-test-kit. As a result of the command actually creating indexes now, CI started failing on QE tests.

https://github.com/prisma/prisma-engines/pull/2287/commits/1c74eca0def7a23809a252508f52e0e15d43e416 reverts the test setup for the QE to the old behaviour, but I imagine we would like to run the QE tests _with indexes_ eventually, so I am creating this issue.

Steps to take:

- Revert the changes in 1c74eca0def7a23809a252508f52e0e15d43e416 and use `SchemaPush` again
- Fix the failing tests. Hopefully, these are just different ordering problems because of the former lack of indexes.

**Comment by matthewmueller** [comment]
For Matt: Needs a discussion/decision

**Comment by matthewmueller** [comment]
My understanding of the current state:

A unique on an optional field will only allow one instance of an undefined field before throwing a unique constraint violation. This differs from SQL databases, where you can have as many NULL values as you want. They don't count towards the uniqueness of the table. 

```prisma
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["mongoDb", "dataProxy"]
}

datasource db {
  provider = "mongodb"
  url      = env("DATABASE_URL")
}

model Log {
  id      String @id @default(dbgenerated()) @map("_id") @db.ObjectId
  level   Level
  message String? @unique
  meta    Json
}
```

For example, the following will throw a unique constraint violation:

```ts
import { PrismaClient } from "@prisma/client"
const prisma = new PrismaClient()

async function main() {
  await prisma.log.create({
    data: {
      level: "Warn",
      meta: {},
    },
  })
  await prisma.log.create({
    data: {
      level: "Warn",
      meta: {},
    },
  })
}

main().catch(console.error)
```

```
PrismaClientKnownRequestError: 
Invalid `prisma.log.create()` invocation in
/Users/m/dev/src/github.com/prisma/ra/mongodb-unique/index.ts:11:20

   8     meta: {},
   9   },
  10 })
→ 11 await prisma.log.create(
  Unique constraint failed on the constraint: `Log_message_key`
    at cb (/Users/m/dev/src/github.com/prisma/ra/mongodb-unique/node_modules/@prisma/client/runtime/index.js:38683:17) {
  code: 'P2002',
  clientVersion: '3.7.0-dev.51',
  meta: { target: 'Log_message_key' }
}
```

And this is currently breaking some QE tests.

## Possible Options

**1. Accept it** document this difference between MongoDB and SQL. It doesn't seem like one is better or worse than the other.

**2. Disable `@unique` on optional fields for MongoDB** prevent this from occurring.

**Comment by matthewmueller** [comment]
^ Candidated this to come to a conclusion. Also not sure what's going on with these tests. Are we skipping them for now?

**Comment by pimeys** [comment]
So, the behavior is the same in MongoDB and SQL Server databases. We made the decision early on in the SQL Server implementation to not do any magic tricks behind the scenes to make all the databases work exactly the same in Prisma. Originally we added a custom partial index to skip using nulls in unique indices, and after a few discussions decided against it. For brownfield projects, if the user already implements a partial index for a unique constraint, our own magic index would clash with the existing setup and make it hard for our users to understand what went wrong when problems appear.

Also the users should expect the databases to work as defined in the manual, which has been a driving factor to almost all of our decisions since Prisma 2.0.

We want to implement a way to define partial indices in the PSL, and we have some issues already (but not for MongoDB):

- https://github.com/prisma/prisma/issues/7263
- https://github.com/prisma/prisma/issues/3076
- https://github.com/prisma/prisma/issues/3387

The problem with partial index is you can define a query what is in the index. For this case you'd define `WHERE foo IS NOT NULL` or something similar, which includes only the not null values in the index (therefore allowing more than one null value). Sadly in many databases this definition allows an arbitrary query, meaning defining all possible cases in the PSL and introspecting them correctly is going to be tough.

**Comment by matthewmueller** [comment]
Thanks @pimeys! I didn't realize SQL Server was the same. Thanks for sharing the prior decision and the principle behind it. 

I feel better going with **1. Accept it**. @dpetrick mentioned that we're already skipping tests, so I think there's nothing left to do other the document the differences. Here's the docs issue: https://github.com/prisma/docs/issues/2695
